ws-server/src/server.rs

```diff
@@ -91,64 +95,114 @@ impl Server {
-							log::warn!("Too many connections. Try again in a while");
-							continue;
-						}
-
-						let conn_counter2 = conn_counter.clone();
-						let shutdown2 = shutdown.clone();
-						let methods = methods.clone();
-						let cfg = self.cfg.clone();
-						let stop_handle2 = self.stop_handle.clone();
-
-						tokio::spawn(async move {
-							let _ = background_task(socket, id, methods, cfg, shutdown2, stop_handle2).await;
-							drop(conn_counter2);
-						});
-
-						id = id.wrapping_add(1);
-					} else {
-						break;
+			match Pin::new(&mut driver).await {
+				Ok((socket, _addr)) => {
+					if let Err(e) = socket.set_nodelay(true) {
+						log::error!("Could not set NODELAY on socket: {:?}", e);
+						continue;
 					}
-				},
-				stop = stop_receiver.next() => {
-					if stop.is_some() {
-						break;
+
+					if driver.connection_count() >= self.cfg.max_connections as usize {
+						log::warn!("Too many connections. Try again in a while.");
+						continue;
 					}
-				},
-				complete => break,
+
+					let methods = &methods;
+					let cfg = &self.cfg;
+
+					driver.add(Box::pin(handshake(socket, id, methods, cfg, &shutdown, &self.stop_handle)));
+
+					id += 1;
+				}
+				Err(DriverError::Io(err)) => {
+					log::error!("Error while awaiting a new connection: {:?}", err);
+				}
+				Err(DriverError::Shutdown) => break,
 			}
 		}
 	}
 }
 
-async fn background_task(
+/// This is a glorified select `Future` that will attempt to drive all
+/// connection futures `F` to completion on each `poll`, while also
+/// handling incoming connections.
+struct ConnDriver<F> {
+	listener: TcpListener,
+	stop_receiver: mpsc::Receiver<()>,
+	connections: Vec<F>,
+}
+
+impl<F> ConnDriver<F>
+where
+	F: Future + Unpin,
+{
+	fn new(listener: TcpListener, stop_receiver: mpsc::Receiver<()>) -> Self {
+		ConnDriver { listener, stop_receiver, connections: Vec::new() }
+	}
+
+	fn connection_count(&self) -> usize {
+		self.connections.len()
+	}
+
+	fn add(&mut self, conn: F) {
+		self.connections.push(conn);
+	}
+}
+
+enum DriverError {
+	Shutdown,
+	Io(std::io::Error),
+}
+
+impl<F> Future for ConnDriver<F>
+where
+	F: Future + Unpin,
+{
+	type Output = Result<(TcpStream, SocketAddr), DriverError>;
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+		let this = Pin::into_inner(self);
+
+		let mut i = 0;
+
+		while i < this.connections.len() {
+			if this.connections[i].poll_unpin(cx).is_ready() {
+				// Using `swap_remove` since we don't care about ordering
+				// but we do care about removing being `O(1)`.
+				//
+				// We don't increment `i` in this branch, since we now
+				// have a shorter length, and potentially a new value at
+				// current index
+				this.connections.swap_remove(i);
+			} else {
+				i += 1;
+			}
+		}
+
+		if let Poll::Ready(Some(())) = this.stop_receiver.next().poll_unpin(cx) {
+			return Poll::Ready(Err(DriverError::Shutdown));
+		}
```

Comment:
This bit replaces the select that was added on `master`.

Reply:
It would be good to document how the shutdown process works and what connected clients should expect to happen to their pending requests.

Reply:
So I'll try again: is **this** where we should send a `CLOSE` frame?

Reply:
Yeah, I reckon @niklasad1's refactor will make the code much easier to follow here too.